includes/utilities.php

```diff
@@ -204,3 +243,49 @@ function add_leading_slash( $string ) {
 
 	return '/' . $string;
 }
+
+/**
+ * Update an option with query parameters that should be preserved for
+ * matching redirects.
+ *
+ * @param array $parameters New query string parameters to add to the preserved list.
+ * @return void
+ */
+function add_preserved_query_string_parameters( array $parameters ) {
+	$existing_parameters = get_preserved_query_parameters();
+	$parameters = array_merge( $existing_parameters, $parameters );
+	$parameters = array_unique( $parameters );
+	$parameters = array_filter( $parameters );
+	update_option( PRESERVED_PARAMS_OPTION, implode( ';', $parameters ) );
+}
+
+/**
+ * Return list of query parameters to preserve.
+ *
+ * @return array
+ */
+function get_preserved_query_parameters() : array {
+	return explode( ';', get_option( PRESERVED_PARAMS_OPTION, '' ) );
+}
+
+/**
+ * Splits query string parameters into preserved and unknown groups.
+ *
+ * Usage: list( $safe, $unknown ) = split_query_parameters( $_SERVER['QUERY_STRING'] );
+ *
+ * @param string $query_string
+ * @return array Array containing 2 associative arrays of preserved and unknown query string parameters.
+ */
+function split_query_parameters( string $query_string ) : array {
```

Comment:
Is this going to handle things like `[]` / `[0]` etc? I'm a but dubious about implementing our own query parsing. Could we just use `$_GET`? Also, I wanted to double check on double-encode issues, has that been tested? i.e. that encoded params are preserved etc.

Reply:
`$_GET` isn't necessarily populated when this is used e.g. in the sanitise and normalise function used when saving a post.

I'll add some more tests to check the encoding. `parse_str()` does handle `[]` etc though again will some additional tests.